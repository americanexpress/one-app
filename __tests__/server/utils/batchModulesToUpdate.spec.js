/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import batchModulesToUpdate from '../../../src/server/utils/batchModulesToUpdate';

jest.mock('../../../src/server/utils/stateConfig', () => ({
  getServerStateConfig: jest.fn(() => ({ rootModuleName: 'my-root' })),
}));

describe('batchModulesToUpdate', () => {
  beforeEach(() => jest.clearAllMocks());

  const modules = [...new Array(65)].map((x, i) => `module-${i}`);
  const modulesWithoutRoot = [...modules];
  modules.splice(20, 0, 'my-root');

  it('should put the root module first', () => {
    const result = batchModulesToUpdate(modules);
    expect(result[0]).toEqual(['my-root']);
  });

  it('should batch the rest of the modules in tens', () => {
    expect.assertions(10);
    const result = batchModulesToUpdate(modules);
    expect(result.length).toBe(8);
    result.forEach((arr, i) => {
      if (i === 0) expect(arr.length).toBe(1);
      else if (i < 7) expect(arr.length).toBe(10);
      else if (i === 7) expect(arr.length).toBe(5);
    });
    expect(JSON.stringify(result)).toMatchSnapshot();
  });

  it('should handle the root module only', () => {
    const result = batchModulesToUpdate(['some-root']);
    expect(result).toEqual([['some-root']]);
  });

  it('should batch modules in tens when there is no root module', () => {
    expect.assertions(9);
    const result = batchModulesToUpdate(modulesWithoutRoot);
    expect(result.length).toBe(7);
    result.forEach((arr, i) => {
      if (i < 6) expect(arr.length).toBe(10);
      else if (i === 6) expect(arr.length).toBe(5);
    });
    expect(JSON.stringify(result)).toMatchSnapshot();
  });
});
