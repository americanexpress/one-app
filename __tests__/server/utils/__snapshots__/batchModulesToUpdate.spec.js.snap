// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`batchModulesToUpdate should batch modules in tens when there is no root module 1`] = `"[["module-0","module-1","module-2","module-3","module-4","module-5","module-6","module-7","module-8","module-9"],["module-10","module-11","module-12","module-13","module-14","module-15","module-16","module-17","module-18","module-19"],["module-20","module-21","module-22","module-23","module-24","module-25","module-26","module-27","module-28","module-29"],["module-30","module-31","module-32","module-33","module-34","module-35","module-36","module-37","module-38","module-39"],["module-40","module-41","module-42","module-43","module-44","module-45","module-46","module-47","module-48","module-49"],["module-50","module-51","module-52","module-53","module-54","module-55","module-56","module-57","module-58","module-59"],["module-60","module-61","module-62","module-63","module-64"]]"`;

exports[`batchModulesToUpdate should batch the rest of the modules in tens 1`] = `"[["my-root"],["module-0","module-1","module-2","module-3","module-4","module-5","module-6","module-7","module-8","module-9"],["module-10","module-11","module-12","module-13","module-14","module-15","module-16","module-17","module-18","module-19"],["module-20","module-21","module-22","module-23","module-24","module-25","module-26","module-27","module-28","module-29"],["module-30","module-31","module-32","module-33","module-34","module-35","module-36","module-37","module-38","module-39"],["module-40","module-41","module-42","module-43","module-44","module-45","module-46","module-47","module-48","module-49"],["module-50","module-51","module-52","module-53","module-54","module-55","module-56","module-57","module-58","module-59"],["module-60","module-61","module-62","module-63","module-64"]]"`;
